### ISSUE TITLE
Confused by README wrt supporting HuggingFace models

### ISSUE THREAD

**Issue by eric-mitchell** [microsoft/LoRA#17]
Thanks for the nice repo! Currently, the readme states:

    This repo contains the source code of the Python package loralib and several examples of how to integrate it with PyTorch models, such as those in HuggingFace.

However, from the examples, it seems like in order to use `loralib` with a HuggingFace model, we need to actually change the code implementation of each model, replacing each nn.Linear with the lora equivalent. If this is the case, I think it's a bit confusing to say the examples show integration with HuggingFace, because as far as I can tell, the examples use re-implementation of GPT2. I was hoping there may be some mechanism to do this automatically, e.g.

    import transformers, loralib
    
    model = transformers.AutoModel.from_pretrained("gpt2")
    lora_model = loralib.wrap(model)  # wrap all nn.Linear modules
    lora_params = loralib.get_lora_only_params(lora_model)

Is this possible? Thanks a lot!

**Comment by eric-mitchell** [comment]
Here is a quick proof of concept that I think will make it easier to add LoRA to any existing model without changing any model code, using the nice [higher](https://github.com/facebookresearch/higher) library:

```
import torch
import torch.nn as nn
from torch.nn import Parameter
from higher.patch import monkeypatch as make_functional


class LoraModel(nn.Module):
    def __init__(self, model, r):
        super().__init__()

        self.base_model = model

        self.lora_params = []
        for p in self.base_model.parameters():
            if p.dim() == 2:
                self.lora_params.append((Parameter(torch.zeros(p.shape[0], r)), Parameter(torch.randn(r, p.shape[1]))))
            else:
                self.lora_params.append(None)

    def forward(self, *args, **kwargs):
        fmodel = make_functional(self.base_model)
        new_params = [p if lp is None else p + lp[0] @ lp[1] for p, lp in zip(self.base_model.parameters(), self.lora_params)]
        fmodel.update_params(new_params)

        return fmodel(*args, **kwargs)

    def lora_parameters(self):
        return [p for lp in self.lora_params if lp is not None for p in lp]


if __name__ == "__main__":
    import transformers

    m = transformers.AutoModelForSequenceClassification.from_pretrained('distilbert-base-cased')
    tok = transformers.AutoTokenizer.from_pretrained('distilbert-base-cased')
    inp = tok("I am a small dog.", return_tensors="pt")
    lm = LoraModel(m, 10)
    output = m(**inp)
    output2 = lm(**inp)
    lora_grads = torch.autograd.grad(output2.logits[0,0], lm.lora_parameters())
    import pdb; pdb.set_trace()

```

Note that this approach doesn't include any fanciness for multi-headed weight matrices. Adding those things probably isn't too hard, though.

**Comment by edwardjhu** [comment]
Hi Eric,

Interesting idea!

By "showing how to integrate" we indeed meant manually replacing some nn.Linear with our implementations, though there shouldn't be too many lines to modify. One consideration is that, in practice, we don't actually want to replace all nn.Linear with lora.Linear. For example, we only replaced the layers corresponding to query and value projection in many of our experiments and obtained satisfactory performance.

I'd love to learn more about if you find it helpful to augmenting all nn.Linear with low-rank matrices, as your code snippet does. 

P.S. My apology for the delayed response! I left Microsoft last year and stopped receiving notifications.

**Comment by eric-mitchell** [comment]
No worries at all. I think the main inconvenience is that many people will be using pre-built models from libraries like HuggingFace, so manually changing lines of library-level code to add LoRA post-hoc adds some non-trivial overhead for quick experiments. I thought maybe clarifying this in the README would be helpful.

I've used the above code a bit in some experiments, and it works well enough, but I haven't directly compared with applying the low-rank matrices at only query/value projections. One could pass a regex to the constructor that matches parameter names from `model.named_parameters()` that should be replaced to get the filtering behavior you're looking for. Anyhow, nice library & paper! Will go ahead and close this.

**Comment by edwardjhu** [comment]
Right, I agree that it would be awesome to integrate with a one-liner. Will note that by "integration" we meant something more involved than using a high-level wrapper.

We thought about the regex approach but decided that it was a bit brittle since people name their modules differently.

Glad to hear that adding low-rank weights to all layers worked for you!

**Comment by yifan1130** [comment]
> No worries at all. I think the main inconvenience is that many people will be using pre-built models from libraries like HuggingFace, so manually changing lines of library-level code to add LoRA post-hoc adds some non-trivial overhead for quick experiments. I thought maybe clarifying this in the README would be helpful.
> 
> I've used the above code a bit in some experiments, and it works well enough, but I haven't directly compared with applying the low-rank matrices at only query/value projections. One could pass a regex to the constructor that matches parameter names from `model.named_parameters()` that should be replaced to get the filtering behavior you're looking for. Anyhow, nice library & paper! Will go ahead and close this.

Thank you for the good codes you provided. Could you just explain more about how to optmize the lora_params you gave? I couldn't let the lora_params into gradient decent

**Comment by eric-mitchell** [comment]
@yifan1130- sorry for the delayed reply. I imagine you might be having some trouble with my code because I'm using a regular list instead of a nn.ParameterList to store the lora_params. If you make that switch, you should be able to pass wrapped_model.lora_params to your optimizer.

**Comment by yifan1130** [comment]
@eric-mitchell Thank you so much for your reply! I have solved the problem. Btw, I wonder would the code you provided support multi-GPU training? I tried to use it for multi-GPU training but didn't have very successful attempts

**Comment by eric-mitchell** [comment]
Glad you were able to solve the problem- I've not actually tried to use higher in multi-GPU setups, so I'm not sure if the issue you're having is based on higher or my code.

One thing I might suggest is switching from higher to `functorch`, which may be more compatible with native PyTorch functionality and better supported/maintained moving forward. It has a [make_functional](https://pytorch.org/functorch/stable/generated/functorch.make_functional.html?highlight=make_functional#functorch.make_functional) function, similar to higher.

**Comment by yifan1130** [comment]
Thank you very much for your suggestions! I will try to use functorch and
see whether this will support multi-GPU training as you suggested！

Eric Mitchell ***@***.***> 于2022年10月26日周三 16:52写道：

**Comment by corbt** [comment]
I'm attempting to fine-tune a RoBERTa model from huggingface transformers without having to fork the model definition. My goal is to match the architecture from the example at https://github.com/microsoft/LoRA/blob/main/examples/NLU/src/transformers/models/roberta/modeling_roberta.py.

This is the code I added to my training script. It runs, but I just started training so will be a while before I know whether it's training properly. 😬

```python
model = AutoModelForSequenceClassification.from_pretrained('roberta-large', num_labels=1)

print("Injecting LoRA layers...")
for name, module in model.named_modules():
    if name.endswith("attention.self"):
        new_query = lora.Linear(
            module.query.in_features, module.query.out_features, 8, 16
        )
        new_key = lora.Linear(module.key.in_features, module.key.out_features, 8, 16)

        # Copy the weights from the original module to the new one
        new_query.weight = module.query.weight
        new_query.bias = module.query.bias
        new_key.weight = module.key.weight
        new_key.bias = module.key.bias

        module.query = new_query
        module.key = new_key
```

**Comment by luffycodes** [comment]
@corbt thanks for the code snippet. Does it perform similarly to [modeling_roberta.py](https://github.com/microsoft/LoRA/blob/main/examples/NLU/src/transformers/models/roberta/modeling_roberta.py)?
